### PR TITLE
Unify Interface projection with Union

### LINKED ISSUE #2847: Relationships to interface types cause problems if the field is mandatory
**Describe the bug**
Relationships to interface types cause problems if the field is mandatory. The object can be created with a mutation, although a mandatory field was not specified. 
If the just created object is read out, there are also problems, but only if the field is queried, which is actually mandatory.

I think there are two separate issues here. Firstly, the input type is not generated correctly (see screenshot below). And on the other hand a query of a type should lead to an error if a mandatory field is not present

**Type definitions**
````gql
interface Product {
  name: String!
}

type Movie implements Product {
  name: String!
}

type Actor {
  name: String!
  product: Product! @relationship(type: "HAS_PRODUCT", direction: OUT)
}
````
See ActorCreateInput from Introspection:
image
product is not mandatory, but should

**To Reproduce**

1. Run a server with the following code...
2. Execute the following Mutation...
````gql
mutation {
  createActors(input: {
    name: "TestActor"
  }) {
    actors {
      name
    }
  }
}
````
Response: Should not work, but it does
````json
{
  "data": {
    "createActors": {
      "actors": [
        {
          "name": "TestActor"
        }
      ]
    }
  }
}
````
3. Then run the following Query...
Works:
````gql
query {
  actors {
    name
  }
}
````
````json
{
  "data": {
    "actors": [
      {
        "name": "TestActor"
      }
    ]
  }
}
````
Doesn't work:
````gql
query {
  actors {
    name
    product {
      name
    }
  }
}
````
````json
{
  "data": {
    "actors": []
  }
}
````
Why is the response empty but successful? I would rather have expected there to be an error response because there is no value for a required field in the database.

**Expected behavior**
- Input types for interface relationships must be generated correctly
- Query must result in an error if a mandatory field is not present.

**System (please complete the following information):**
* Version: @neo4j/graphql@3.16.0
* Node.js version: 16.16.0

### LINKED ISSUE #2883: Nodes may disappear if an interface is requested that is not connected
I will later provide a more complete example of this.

```GraphQL
interface I {
  foo: String!
}

type A implements I {
  foo: String!
}

type B implements I {
  foo: String!
}

type T {
  bar: Int!
  related: I @relationship(type: "SOMETHING", direction: OUT)
}
```

If we would now have several objects of type T only the ones with a SOMETHING edge would be returned.

For example, this  might return 5 items:

```GraphQL
{
  ts {
    bar
  }
}
```

But this might return only 2 items:

```GraphQL
{
  ts {
    bar
    related {
      foo
    }
  }
}
```

Ts that not have this the relation would be missing in the list. On the other hand, if queried via a Connection it would work. Is this intentional? It seems wrong. The issue only occurs if interfaces are used, not if "related" is directly of type A or B.

### PR DESCRIPTION
**Pull request by MacondoExpress**
# Unify Interface projection with Union

This PR wants to reuse the union projection implementation for interfaces.
As bug #2847 is not present on the union implementation it closes it as well.

## Complexity

Complexity: Mid

# Issue

Closes #2847, closes #2883

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [x] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 32 | 40 | 1134 |
|  aggregations.NestedAggregation | 13514 | 13514 | 94 | 111 | 2174 |
|  aggregations.AggregationWithWhere | 10942 | 10942 | 53 | 71 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 20101954 | 20101954 | 2886 | 2905 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 20101954 | 20101954 | 2648 | 2723 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 8694 | 8694 | 48 | 61 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 8900 | 8900 | 50 | 58 | 2174 |
|  aggregations.DeeplyNestedCount | 10054408 | 10054408 | 4261 | 4104 | 2008534 |
|  batch-create.BatchCreate | 3600 | 3600 | 97 | 119 | 600 |
|  connect.createAndConnect | 12419 | 12419 | 200 | 248 | 3003 |
|  connections.Connection | 13042 | 13042 | 118 | 116 | 2174 |
|  connections.NestedConnection | 33883 | 33883 | 98 | 123 | 4516 |
|  connections.ConnectionWithSort | 2277 | 2277 | 56 | 77 | 1040 |
|  connections.ConnectionWithSortAndCypher | 2277 | 2277 | 155 | 309 | 1040 |
|  create.SimpleMutation | 6 | 6 | 30 | 39 | 1 |
|  cypher-directive.TopLevelCypherDirective | 99525 | 99525 | 768 | 710 | 10168 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 127992 | 127994 | 295 | 258 | 10168 |
|  delete.SimpleDelete | 18361 | 18361 | 299 | 283 | 1040 |
|  delete.NestedDeleteInUpdate | 16779 | 16779 | 119 | 321 | 2040 |
|  query.SimpleQuery | 14120 | 14120 | 181 | 92 | 2174 |
|  query.QueryWhere | 8681 | 8656 | 32 | 50 | 2168 |
|  query.SimpleQueryWithNestedWhere | 8852 | 8827 | 52 | 70 | 2168 |
|  query.Nested | 10084290 | 10084290 | 9395 | 9760 | 2008534 |
|  query.NestedWithFilter | 37172 | 37172 | 120 | 116 | 4511 |
|  query.OrFilterOnRelationships | 35829 | 36723 | 157 | 257 | 1635 |
|  query.OrFilterOnRelationshipsAndNested | 29700 | 30178 | 193 | 266 | 1635 |
|  query.QueryWithNestedIn | 13317 | 12323 | 55 | 60 | 1540 |
|  query.NestedConnectionWhere | 8794 | 8794 | 65 | 76 | 2174 |
|  query.DeeplyNestedConnectionWhere | 8869 | 8808 | 74 | 112 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 6251 | 6251 | 131 | 165 | 1134 |
|  query.NestedWithRelationshipSingleFilters | 3021189 | 3021189 | 557 | 651 | 1003942 |
|  query.Fulltext | 96 | 96 | 24 | 30 | 16 |
|  query.FulltextWithNestedQuery | 571 | 571 | 38 | 42 | 84 |
|  sorting.SortMultipleTypes | 2493 | 2493 | 59 | 85 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1430 | 1446 | 182 | 212 | 1040 |
|  sorting.SortOnNestedFields | 13042 | 13042 | 137 | 73 | 2174 |
|  sorting.SortDeeplyNestedFields | 38965 | 38965 | 196 | 186 | 4516 |
|  sorting.SortWithTopLevelCypher | 2119 | 2119 | 90 | 156 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 57 | 70 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 59 | 77 | 35 |
|  unions.NestedUnion | 398553 | 398553 | 327 | 303 | 33033 |
|  unions.NestedUnionWithMissingFields | 372527 | 372527 | 273 | 315 | 33033 |
|  update.NestedUpdate | 2119 | 2119 | 36 | 40 | 1040 |

</details>
        
Old Schema Generation: 27.565s
Schema Generation: 27.892s

File: packages/graphql/src/translate/create-projection-and-params.ts [hunk 1]
```diff
@@ -470,3 +472,26 @@ function createFulltextProjection({
...
+            } else {
+                nodeWhere = {
+                    [sharedFilterKey]: sharedFilterValue,
+                };
+            }
+            where["_on"][node.name] = nodeWhere;
+        });
+    });
+    return where;
+}
```
Comment (darrellwarde) [inline packages/graphql/src/translate/create-projection-and-params.ts:497]: This is the kind of function that I would really start to see unit tests for - I really don't have 100% confidence in what this is doing just by reading it alone.

**Review by darrellwarde — APPROVED** [review]
This seems sound to me - not the easiest code in the world to understand, we should look at improving that.

**Comment by MacondoExpress** [comment]
I 100% agree with you, this code put the interface filters into the implementations filters, so if you have a `Production` interface and a `Movie` and `Series` implementation and you apply a filter on the production interface this code transform this:

```graphql
   query Actors{
                actors(where: { name: "Ashton" }) {
                    name
                    actedIn(where: { title: "The 70's show", _on: { Movie: { title: "A Movie" } } }) {
                        title
                        ... on Movie {
                            runtime
                        }
                        ... on Series {
                            episodes
                        }
                    }
                }
            }
```
to 

```graphql
query Actors{
                actors(where: { name: "Ashton" }) {
                    name
                    actedIn(where: { _on: { Movie: { title: "A Movie" }, Series:  { title: "The 70's show" } } }) {
                        title
                        ... on Movie {
                            runtime
                        }
                        ... on Series {
                            episodes
                        }
                    }
                }
            }
```
The order in which is executed is important to define which implementations we want to return.

It will be completely removed by the parser to generate the Composite Entity with the IR project.

I'll maybe simplify the implementation a bit before merging it as maybe the parser will take time to develop!

**Review by a-alle — APPROVED** [review]

**Pull request merged**